utils/google.js

```diff
@@ -254,4 +309,165 @@ module.exports = {
+
+    console.log(`Trying to lock Google project "${googleProject.id}"...`);
+
+    /**
+     * return true if the project was successfully locked, false otherwise
+     */
+    const tryLockProject = async function(googleProject) {
+      // check if the project is locked by anyone (not using the unique name)
+      listRes = await module.exports.listServiceAccounts(googleProject.id);
+      isLocked = listRes.some(sa => sa.email.startsWith(lockServiceAccountName));
+      if (isLocked) return false;
+
+      const createRes = await module.exports.createServiceAccount(googleProject.id, lockName, 'Locked: a test is currently using the project');
+
+      if (typeof createRes === 'object' && createRes instanceof Error) {
+        console.log('Unable to lock the project even though it is not already locked...');
+        console.log(createRes);
+        return false;
+      } else {
+        expect(createRes).to.have.property('email');
+        expect(createRes.email).to.equal(serviceAccountEmail);
+        console.log('Successully locked project');
+        return true;
+      }
+    };
+
+    try {
+      await apiUtil.smartWait(tryLockProject, [googleProject], timeout, `Could not lock project after ${timeout} secs`);
+    }
+    catch(err) {
+      console.log(err);
+      return false;
+    }
+
+    // It sometimes takes a bit of time for the lock to be in the list of SAs
+    // We need it be be listed to make sure another session cannot lock
+    /**
+     * return true if the locking SA is in the list, false otherwise
+     */
+    let isLockingSaInList = async function() {
+      let listRes = await module.exports.listServiceAccounts(googleProject.id);
+      return listRes.some(sa => sa.email === serviceAccountEmail);
+    };
+    let listTimeout = 30;
+    try {
+      await apiUtil.smartWait(isLockingSaInList, [], listTimeout, `Registered lock SA is not listed after ${listTimeout} secs`, startWait=1);
+      return true;
+    }
+    catch(err) {
+      console.log(err);
+      return false;
+    }
+  },
+
+  /**
+   * Unlocks a Google project
+   * Returns true if nothing to unlock or successfully unlocked, false
+   * otherwise
+   */
+  async unlockGoogleProject(googleProject) {
+    const serviceAccountEmail = this.getLockingServiceAccountEmail(googleProject.serviceAccountEmail);
+
+    // check if the project is locked. if not locked: return success
+    listRes = await this.listServiceAccounts(googleProject.id);
+    isLocked = listRes.some(sa => sa.email.startsWith(lockServiceAccountName));
+    if (!isLocked) return true;
+
+    // if it's locked by another testing session: cannot unlock
+    isLockedByMe = listRes.some(sa => sa.email === serviceAccountEmail);
+    if (!isLockedByMe) return false;
+
+    // if it's lock by this testing session: try to unlock
+    console.log(`Unlocking Google project "${googleProject.id}"`);
+    const deleteRes = await this.deleteServiceAccount(serviceAccountEmail, googleProject.id);
```

Comment:
does this work pretty much every time or should you retry this if it fails?

Reply:
i have never seen it fail